src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+
+  /// \brief Default constructor; constructs an empty sparse matrix.
+  ///
+  /// FIXME (mfh 09 Aug 2013) numCols and nnz should be properties of
+  /// the graph, not the matrix.  Then BlockCrsMatrix needs methods to get
+  /// these from the graph.
+  BlockCrsMatrix () :
+    numCols_ (0),
+    blockDim_ (0)
+  {}
+
+  //! Copy constructor (shallow copy).
+  template<typename SType,
+           typename OType,
+           class DType,
+           class MTType,
+           typename IType>
+  BlockCrsMatrix (const BlockCrsMatrix<SType,OType,DType,MTType,IType> & B) :
+    graph (B.graph.entries, B.graph.row_map),
+    values (B.values),
+    dev_config (B.dev_config),
+#ifdef KOKKOS_USE_CUSPARSE
+    cusparse_handle (B.cusparse_handle),
+    cusparse_descr (B.cusparse_descr),
+#endif // KOKKOS_USE_CUSPARSE
+    numCols_ (B.numCols ()),
+    blockDim_ (B.blockDim ())
+  {
+    graph.row_block_offsets = B.graph.row_block_offsets;
+    //TODO: MD 07/2017: Changed the copy constructor of graph
+    //as the constructor of StaticCrsGraph does not allow copy from non const version.
+  }
+
+  /// \brief Construct with a graph that will be shared.
+  ///
+  /// Allocate the values array for subsquent fill.
+  BlockCrsMatrix (const std::string& arg_label,
+             const StaticCrsGraphType& arg_graph, const OrdinalType& blockDim) :
+    graph (arg_graph),
+    values (arg_label, arg_graph.entries.extent(0)),
+    numCols_ (maximum_entry (arg_graph) + 1),
+    blockDim_ (blockDim)
+  {}
+
+  /// \brief Constructor that copies raw arrays of host data in
+  ///   coordinate format.
+  ///
+  /// On input, each entry of the sparse matrix is stored in val[k],
+  /// with row index rows[k] and column index cols[k].  We assume that
+  /// the entries are sorted in increasing order by row index.
+  ///
+  /// This constructor is mainly useful for benchmarking or for
+  /// reading the sparse matrix's data from a file.
+  ///
+  /// \param label [in] The sparse matrix's label.
+  /// \param nrows [in] The number of rows.
+  /// \param ncols [in] The number of columns.
+  /// \param annz [in] The number of entries.
+  /// \param val [in] The entries.
+  /// \param rows [in] The row indices.  rows[k] is the row index of
+  ///   val[k].
+  /// \param cols [in] The column indices.  cols[k] is the column
+  ///   index of val[k].
+  /// \param pad [in] If true, pad the sparse matrix's storage with
+  ///   zeros in order to improve cache alignment and / or
+  ///   vectorization.
+  ///
+  /// FIXME (mfh 21 Jun 2013) The \c pad argument is currently not used.
+  BlockCrsMatrix (const std::string &label,
+             OrdinalType nrows,
+             OrdinalType ncols,
+             size_type annz,
+             ScalarType* val,
+             OrdinalType* rows,
+             OrdinalType* cols,
+             OrdinalType blockdim,
+             bool pad = false)
+  {
+    (void) pad;
+    import (label, nrows, ncols, annz, val, rows, cols, blockdim);
```

Comment:
This is an unfortunate name because it collides with both Epetra and Tpetra in confusing ways.

Reply:
Yes, responded to comment below to see if this can be removed (or at least renamed and hidden).